commit 2b0f9cf9e0e0a231a30d6273466d42eb9da56a33
Author: Yang Bo <yangbo@linuxdeepin.com>
Date:   Sun Jan 3 20:45:41 2016 +0800

    EDAC, sb_edac: Fix TAD presence check for sbridge_mci_bind_devs()
    
    BugLink: http://bugs.launchpad.net/bugs/1514889
    
    (commit 2900ea609616c2651dec65312beeb2a6e536bc50 upstream.)
    
    In commit
    
      7d375bffa524 ("sb_edac: Fix support for systems with two home agents per socket")
    
    NUM_CHANNELS was changed to 8 and the channel space was renumerated to
    handle EN, EP, and EX configurations.
    
    The *_mci_bind_devs() functions - except for sbridge_mci_bind_devs() -
    got a new device presence check in the form of saw_chan_mask. However,
    sbridge_mci_bind_devs() still uses the NUM_CHANNELS for loop.
    
    With the increase in NUM_CHANNELS, this loop fails at index 4 since
    SB only has 4 TADs.  This results in the following error on SB machines:
    
      EDAC sbridge: Some needed devices are missing
      EDAC sbridge: Couldn't find mci handler
      EDAC sbridge: Couldn't find mci handle
    
    This patch adapts the saw_chan_mask logic for sbridge_mci_bind_devs() as
    well.
    
    After this patch:
    
      EDAC MC0: Giving out device to module sbridge_edac.c controller Sandy Bridge Socket#0: DEV 0000:3f:0e.0 (POLLED)
      EDAC MC1: Giving out device to module sbridge_edac.c controller Sandy Bridge Socket#1: DEV 0000:7f:0e.0 (POLLED)
    
    Signed-off-by: Seth Jennings <sjenning@redhat.com>
    Acked-by: Aristeu Rozanski <aris@redhat.com>
    Acked-by: Tony Luck <tony.luck@intel.com>
    Tested-by: Borislav Petkov <bp@suse.de>
    Cc: Mauro Carvalho Chehab <mchehab@osg.samsung.com>
    Cc: linux-edac <linux-edac@vger.kernel.org>
    Link: http://lkml.kernel.org/r/1438798561-10180-1-git-send-email-sjenning@redhat.com
    Signed-off-by: Borislav Petkov <bp@suse.de>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
    
    (cherry picked from commit 362e91b58e621e61ebc05c0864b2cf45e91dd6e7)
    Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
    Signed-off-by: Luis Henriques <luis.henriques@canonical.com>
    Signed-off-by: Yang Bo <yangbo@linuxdeepin.com>

diff --git a/drivers/edac/sb_edac.c b/drivers/edac/sb_edac.c
index cf1268d..429309c 100644
--- a/drivers/edac/sb_edac.c
+++ b/drivers/edac/sb_edac.c
@@ -1688,6 +1688,7 @@ static int sbridge_mci_bind_devs(struct mem_ctl_info *mci,
 {
 	struct sbridge_pvt *pvt = mci->pvt_info;
 	struct pci_dev *pdev;
+	u8 saw_chan_mask = 0;
 	int i;
 
 	for (i = 0; i < sbridge_dev->n_devs; i++) {
@@ -1721,6 +1722,7 @@ static int sbridge_mci_bind_devs(struct mem_ctl_info *mci,
 		{
 			int id = pdev->device - PCI_DEVICE_ID_INTEL_SBRIDGE_IMC_TAD0;
 			pvt->pci_tad[id] = pdev;
+			saw_chan_mask |= 1 << id;
 		}
 			break;
 		case PCI_DEVICE_ID_INTEL_SBRIDGE_IMC_DDRIO:
@@ -1741,10 +1743,8 @@ static int sbridge_mci_bind_devs(struct mem_ctl_info *mci,
 	    !pvt-> pci_tad || !pvt->pci_ras  || !pvt->pci_ta)
 		goto enodev;
 
-	for (i = 0; i < NUM_CHANNELS; i++) {
-		if (!pvt->pci_tad[i])
-			goto enodev;
-	}
+	if (saw_chan_mask != 0x0f)
+		goto enodev;
 	return 0;
 
 enodev:
